_posts/2017/2017-06-02-reactive-streams.md

```diff
@@ -0,0 +1,149 @@
+
+* `Source[Out, M]`: In Akka streams a source is a stream stage that only has a downstream side.
+  So basically it creates elements out of thin air.
+  For example by an algorithm or by reading data from a disk (the data from the disk is not thin air, but from an API point of view there is no further upstream stage).
+* `Flow[In, Out, M]`: A flow is a stream stage that has an upstream and a downstream side.
+  So elements flow in are processed and then passed downstream.
+* `Sink[In, M]`: A sink is a stream stage that only has an upstream side.
+  So it has elements flowing in but nothing flowing out.
+  This could be console logging or writing to disk.
+
+Here is the example code:
+
+{% highlight scala %}
+import java.nio.file.Paths
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model.{HttpMethods, HttpRequest, HttpResponse}
+import akka.stream.scaladsl.{FileIO, Flow, Sink, Source}
+import akka.stream.{ActorMaterializer, ThrottleMode}
+import akka.util.ByteString
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+object AkkaStreamsExample {
+  implicit val system = ActorSystem()
+  implicit val materializer = ActorMaterializer()
+
+  // a future on the http
+  val httpResponse: Future[HttpResponse] = Http().singleRequest(HttpRequest(HttpMethods.GET, "http://my.domain.com/file.tar.gz"))
+  // convert the future into a source that emits a single element (the response)
+  val httpResponseSource: Source[Source[ByteString, Any], NotUsed] = Source.fromFuture(httpResponse.map(_.entity.dataBytes))
+  // convert the single response source into a source of the response body byte chunks
+  val httpResponseBodySource: Source[ByteString, NotUsed] = httpResponseSource.flatMapConcat(identity)
+
+  val clientWebbrowser: Flow[ByteString, ByteString, NotUsed] =
+    Flow[ByteString]
+      .map { chunk =>
+        println(s"Received ${chunk.length} bytes")
+        // browser does nothing, just passes down bytes chunk unaltered
+        chunk
+      }
+
+  val clientDisk: Sink[ByteString, NotUsed] =
+    Flow[ByteString]
+      // throttle to at most 3 byte chunks per second to simulate a slow disk
+      // to throttle down to a certain bytes/second speed, we would have to also incorporate
+      // the length of every byte chunk, not only the chunk count
+      .throttle(3, 1.second, 3, ThrottleMode.shaping)
+      // all inbound byte chunks get written to disk
+      .to(FileIO.toPath(Paths.get("/path/to/target/file")))
+
+  // download the file
+  httpResponseBodySource.via(clientWebbrowser).to(clientDisk).run()
+}
+{% endhighlight %}
+
+Beautiful just by itself, isn't it? And it is light on used resources.
+We built it to be artifically very slow.
+Most of the time is waiting because of the throttling.
+Still there is no thread bound to all this.
+We only dedicate a thread to do out work (the `println` part) when there is work to do.
```

Comment:
do you mean outside work in terms of putting out sth. or using the thread at all? Sentence is hard to understand